src/server/auth.ts

```diff
@@ -63,6 +65,22 @@ export const authOptions: NextAuthOptions = {
           return null;
         }
 
+        const user = await db.users.findUnique({
+          where: { email: c.data.email },
+        });
+
+        if (
+          user &&
+          user.deprecatedPasswordDigest &&
+          bcrypt.compareSync(
```

Comment:
Note: Notice that on bcrypt's [gh doc](https://github.com/kelektiv/node.bcrypt.js?tab=readme-ov-file#why-is-async-mode-recommended-over-sync-mode) it is suggested that:

"We recommend using async API if you use bcrypt on a server. Bcrypt hashing is CPU intensive which will cause the sync APIs to block the event loop and prevent your application from servicing any inbound requests or events. The async version uses a thread pool which does not block the main event loop"

Maybe @davidlhw could you elaborate the thought process on deciding between `compareSync` & `compare` in this case? Think it will be a good note for everyone to learn 😆

Reply:
for future reference:

the requirement here is that password verifications needs to be done before the user is allowed to view protected resources

if we were to use the async option (without `await` that is, to leverage asynchrony), what essentially happens is that the program would run the `compare` function, and while `compare` is computing, proceed with subsequent computations

in the case where `compare` is used in literal replacement with `compareSync`, the `true && bcrypt.compare(...)` statement will always yield `true` as a promise object is "truthy". This would mean that the user would then be authenticated even if the passwords do not match the user's v1 credentials

we lose benefits of asynchony with an `await` as the rest of the function sleeps while waiting for `compare` to complete. this is similar in behavior with `compareSync`. see [async+await vs sync](https://softwareengineering.stackexchange.com/questions/433640/in-javascript-how-is-awaiting-the-result-of-an-async-different-than-sync-calls)

computation of hash comparison is blocking but only for the currently active function. the application itself will still run fine, especially since nextjs backend api deployed on vercel is built on top of [serverless functions](https://nextjs.org/learn-pages-router/basics/api-routes#:~:text=Next.js%20has%20support%20for%20API%20Routes%2C%20which%20let%20you%20easily%20create%20an%20API%20endpoint%20as%20a%20Node.js%20serverless%20function.), which have it's own "isolated" runtime. hence, other operations will simply invoke new functions and will not prevent the whole application from crashing

feel free to resolve if this makes sense!